src/discovery/index.ts

```diff
@@ -5,13 +5,31 @@ const NODES_SET = 'colyseus:nodes';
 const DISCOVERY_CHANNEL = 'colyseus:nodes:discovery';
 
 export interface Node {
-    port: number;
-    processId: string;
+  port: number;
+  processId: string;
+}
+
+async function getNodeHost() {
+  // allow end users to specify a kubernetes internal cluster IP
+  if (process.env.GET_HOSTNAME_FROM === 'env') {
+    return process.env.SELF_HOSTNAME;
+  }
+
+  // Default to dns resolution
+  return await ip.v4();
+}
+
+function getPort(node: Node) {
+  if (process.env.GET_HOSTNAME_FROM === 'env') {
+    return process.env.SELF_PORT || node.port;
+  }
+
+  return node.port;
 }
 
 async function getNodeAddress(node: Node) {
-  const ipv4 = await ip.v4();
-  return `${node.processId}/${ipv4}:${node.port}`;
+  const host = await getNodeHost();
+  return `${node.processId}/${host}:${getPort(node)}`;
```

Comment:
We could put both of them here, without a separate function for each of them

Reply:
done